toc/rfc/app-runtime-interfaces/rfc-draft-user-creation-by-org-managers.md

```diff
@@ -0,0 +1,63 @@
+# Meta
+[meta]: #meta
+- Name: User Creation by Org Managers
+- Start Date: 2024-08-13
+- Author(s): stephanme
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: [#946](https://github.com/cloudfoundry/community/pull/946)
+
+
+## Summary
+
+Allow Org Managers (and CF Admins) to create users in UAA in order to improve the onboarding procedure for larger developer groups into multi-tenant Cloud Foundry foundations. The users are created by the Cloud Controller on behalf of the Org Manager.
+
+## Problem
+
+In order to assign CF roles to users, the users must be available in the UAA. As of today, the user is created in UAA explicitly by an administrator (origin=uaa) or implicitly when the user authenticates the first time via CF UAA for setups where an external identity provider is used (other origins).
+
+This is an issue for customers that want to onboard larger developer groups into multi-tenant Cloud Foundry foundations. All developers would have to log-in into the foundation before CF roles can be assigned - which is impractical. Customers should be able to automate the complete onboarding e.g. using CF CLI or other CF API clients like a CF terraform provider.
+
+## Proposal
+
+The proposal is to grant Org Managers the right to assign CF roles to users which are not yet known in UAA. The users are created in UAA either explicitly via an enhanced `POST /v3/users` call or implicitly when an Org Manager assigns an organization role using `POST /v3/roles`.
+
+The functionality SHALL be gated by a deployment manifest configuration flag and SHALL be disabled by default.
+
+## Required Changes by Component
+
+### Cloud Controller
+
+#### CAPI Release
+
+Introduce a configuration flag `cc.allow_user_creation_by_org_manager` which is disabled by default .
+
+#### POST /v3/users
+
+Enhance the [POST /v3/users](https://v3-apidocs.cloudfoundry.org/version/3.172.0/index.html#create-a-user) endpoint by the parameters `username` and `origin`. `guid` and `username` parameters are exclusive.
+The existing user creation by `guid` SHALL NOT be changed.
+
+Required Parameters for Create User by Name
+|Name|Type|Description|
+|---|---|---|
+|username|string|The username as (to be) registered in UAA.|
+|origin|string|The origin / identity provider for the user. `origin=uaa` is not allowed as it requires a password. |
+
+
+Permitted roles for Create User by Name
+- Admin
+- OrgManager (if configuration flag `cc.allow_user_creation_by_org_manager` is enabled)
+
+CAPI will call the [POST /Users](https://docs.cloudfoundry.org/api/uaa/version/77.14.0/index.html#create-2) endpoint of the UAA using a CAPI owned UAA client with scope `scim.write` or `uaa.admin` for creating the user if it doesn't exist yet in UAA. The user is then registered in the Cloud Controller database under the user guid returned by UAA.
```

Comment:
To confirm, 
1. This will be a new client, not adding scopes to the [existing usernames lookup client](https://github.com/cloudfoundry/cf-deployment/blob/2bb7d43f751714c8049b8d612a5c9e6836b9f345/cf-deployment.yml#L689)
2. This client will NOT be included by default in cf-deployment

Are those correct?

Reply:
That's correct. We would add a new client e.g. cloud_controller_shadow_user_creation (shadow user because the use case works only for external IDPs).

A new ops file in cf-deployment could set the configuration flag and configure the new UAA client so that the client doesn't exist by default. CC would respond with a 5xx if `cc.allow_user_creation_by_org_manager` is enabled but the UAA client is missing.

Reply:
I mentioned the extra UAA client in the RFC and added a section for a cf-deployment ops file that enables the feature and provisions the additional UAA client.